### DISCUSSION TITLE
How to use `@vitejs/plugin-legacy` correctly?

### DISCUSSION THREAD

**Discussion by ShumRain** [remix-run/remix#8887] [Q&A]
We using the `@vitejs/plugin-legacy` in Remix. This plugin uses `transformIndexHtml` to insert `<script src="legacyPolyfillFilename.js" >` in the HTML. 

When `remix vite:build` is executed using `@vitejs/plugin-legacy`, a `polyfills-xxxx.js` file is generated under `/build/assets/`.

I checked the `buid/.vite/client-manifest.json` file:
```json
{
  ...
  "vite/legacy-polyfills": {
    "file": "assets/polyfills-xxxx.js",
    "src": "vite/legacy-polyfills",
    "isEntry": true
  }
}
```

But `vite/legacy-polyfills` is not processed during the conversion from `buid/.vite/client-manifest.json` to `build/manifest.js`.

How to use `@vitejs/plugin-legacy` correctly?

**Comment by AdiRishi** [comment]
+1 on this, does anyone have a solution yet?

**Reply by ShumRain** [reply]
I have a temporary solution, but it's not recommended for production environments.

I changed some code and used [patch-package](https://github.com/ds300/patch-package) to keep the changes.

```js
// /node_modules/@remix-run/dev/dist/vite/plugin.js
-    await writeFileSafe(path__namespace.join(getClientBuildDirectory(ctx.remixConfig), manifestPath), `window.__remixManifest=${JSON.stringify(manifest)};`);

+    let importPolyfill = ''
+    if (viteManifest['vite/legacy-polyfills']) {
+      const polyfillEntry = resolveBuildAssetPaths(ctx, viteManifest, 'vite/legacy-polyfills')
+
+      if (polyfillEntry.module) {
+        importPolyfill = `import '${polyfillEntry.module}';`
+      }
+    }
+
+    await writeFileSafe(
+      path__namespace.join(getClientBuildDirectory(ctx.remixConfig), manifestPath),
+      `${importPolyfill}window.__remixManifest=${JSON.stringify(manifest)};`
+    );
```

**Reply by AdiRishi** [reply]
Maybe this discussion should be an issue rather than a discussion then. Seems like this is a legit bug.

**Reply by kirbysayshi** [reply]
I just ran into this too. It's not just legacy-polyfills, I also couldn't get the modern polyfills or even legacy bundles to load. So it seems like an incompatibility with the plugin and remix!

**Reply by CMCDragonkai** [reply]
Is this the proper way of introducing automatic polyfills in a remix & vite app? If so, then why is this not part of the documentation?

**Reply by CMCDragonkai** [reply]
I've investigated this a bit.

I think the polyfill file shouldn't really be put into the manifest file.

Normally it would add:

```
    <script type="module" crossorigin src="/assets/polyfills-B1wrU7sj.js"></script>
```

To the top of `<head>`.

Since the generated HTML file would instead have a script section that imports the manifest, I think whatever is generating the HTML should just ALSO conditionally import the polyfill file if it exists.

So the resulting response from the HTTP server should look like:

```
    <script type="module" async="">
        import "/assets/polyfill-....js";
        import "/assets/manifest-15cc1522.js";
        import * as route0 from "/assets/root-B3jGLBi6.js";
        import * as route1 from "/assets/_index-BPE4r6-g.js";

        window.__remixRouteModules = { "root": route0, "routes/_index": route1 };

        import("/assets/entry.client-uPXjd4nb.js");
    </script>

</body>

</html>
```

**Reply by CMCDragonkai** [reply]
Which file controls the output of the final `<script>` tag in remix-run?

**Reply by CMCDragonkai** [reply]
The other issue is that even if it works in production by looking up the polyfill file in the generated files, that doesn't solve the development server problem. Seems like that would require https://github.com/remix-run/remix/discussions/8887#discussioncomment-8848153.

**Reply by tyouzu1** [reply]
> Is this the proper way of introducing automatic polyfills in a remix & vite app? If so, then why is this not part of the documentation?

I think you should give up and embrace [progressive enhancement](https://remix.run/docs/en/main/guides/browser-support),If you look at the source code, you will know that they will not worry about this in the short term, because their focus is on react-router and shopify or whatever.

If you can't accept progressive enhancement for a while, I suggest you use next.js or vike first.But it really depends on your user base.

**Reply by CMCDragonkai** [reply]
My polyfills have nothing to do with the ESM support. It's a web stream API that I'm trying to polyfill, but anyway I got it to work for prod. But not sure how to hack it into the dev server. I think at least remix should just support the legacy plugin and put into their client manifest.
